rust/projects/parallelism/project.md

```diff
@@ -0,0 +1,447 @@
+
+It's nothing complicated. Instead of creating a new thread for every
+multi-threaded job to be performed, a thread pool maintains a "pool" of threads,
+and reuses those threads instead of creating a new one.
+
+But why?
+
+It's entirely about performance. It saves a small amount of performance, but
+when writing high-performance applications, every bit counts. Imagine what it
+takes to make a new thread:
+
+You've got to have a call stack for that thread to run on. That's an allocation.
+Allocations are pretty cheap, but not as cheap as not allocation. How that call
+stack is allocated depends on details of the operating system and runtime, but
+can involve locks and syscalls. Syscalls again are not _that_ expensive, but
+they are expensive when we're dealing with Rust levels of performance. Reducing
+syscalls is a common source of easy optimization. That stack then has to be
+carefully initialized so that first [stack _frame_] contains the appropriate
+values for the return pointer and other things. In Rust the stack needs to be
+configured with a [guard page]. That's two more syscalls, [to `mmap` and to
+`mprotect`][mp].
+
+[2mb]: https://github.com/rust-lang/rust/blob/6635fbed4ca8c65822f99e994735bd1877fb063e/src/libstd/sys/unix/thread.rs#L12
+[mp]: https://github.com/rust-lang/rust/blob/6635fbed4ca8c65822f99e994735bd1877fb063e/src/libstd/sys/unix/thread.rs#L315
+
+TODO: illustration?
+
+That's just setting up the callstack. It's at least another syscall to create
+the new thread, at which point the kernel must do its own accounting for the new
+thread.
+
+In Rust libpthread handles most of this complexity.
+
+Then at some point the OS performs a [context switch] onto the new thread. With
+a thread pool, all that setup overhead is only done for a few threads, and
+subsequent jobs are simply context switches into that pool of threads.
+
+So how do you build a thread pool?
+
+There are many strategies and tradeoffs, but for this exercise you are going to
+write a "round-robin" thread pool. "Round-robin" just means that you have a
+circular list of threads and distribute work to them in order. This is the very
+simplest work scheduling strategy, but it can be effective. What are the
+downsides?
+
+There is one wrinkle in this easy strategy: your thread pool will need to deal
+with the case where the spawned function panics &mdash; simply letting panics
+destroy the threads in your pool would quickly deplete its pool of threads. So
+if a thread in your pool panics you need to make sure that the total number of
+threads doesn't decrease. So what should you do? You have two options: let the
+thread die and spawn another, or catch the panic and keep the existing thread
+running. What are the tradeoffs? You've got to pick one, but leave a comment in
+your code explaining your choice.
+
+_Create the `RoundRobinThreadPool` type, implementing `ThreadPool`_.
+
+The tools at your disposal are `[thread::spawn]`, [`catch_unwind`], [`mpsc`]
```

Comment:
It's OK to use `std::mpsc` to implement a round-robin thread pool. But I just realize that it is extremely easy to use mpmc crossbeam channel to implement a thread pool. Every thread just loops and receives tasks from the channel...

Reply:
That would though not be round-robin, and I'm hoping that round-robin performs particularly badly in comparison to a production threadpool.

I would like to introduce crossbeam though, and it is much easier to write a threadpool with a single queue. Hm.

I'll change it and hope that we can find maybe a work-stealing threadpool that performs better. I have only heard bad things about every rust thread pool... but I'm planning to use rayon and hope for the best.

Reply:
Maybe I'll try to leave the strategy open, have some background reading about scheduling strategies, and point to a few tools.